### PR TITLE
feat(api): allow ProtocolContext to load labware using the ProtocolEngine

### LINKED ISSUE #7390: feat: ProtocolEngineImplementation issues load labware command
## Overview

The `load_labware` method of the newly created `ProtocolEngineImplementation` should execute a `load_labware` command on the Protocol Engine.

The `labware_id` returned by the engine response can be used to construct a `LabwareInterface`, but that functionality is out of scope for this ticket.

This ticket pulls off a chunk of #7325 to reduce its scope and feed some extra clarity into #7332.

## Implementation details

1. Add a new interface to the protocol engine that constructs a LoadLabwareRequest and returns a LoadLabwareResult
    - Eventually, we probably want one of these helper methods for every single command
    - The primary consumer of this interface will be the ProtocolContext
2. Call this method within the `ProtocolEngineImplementation`
3. Continue to raise a `NotImplemented` error, as the correct LabwareInterface has not yet been implemented and will be wired up in #7325  

## Acceptance criteria

- [ ] When combined with #7332, #7325 is unblocked

### LINKED ISSUE #7448: Protocol Engine LabwareGeometry

### PR DESCRIPTION
**Pull request by mcous**
## Overview

PR closes #7390, closes #7448, see tickets for details

## Changelog

1. Added a test to the `ProtocolEngineContext` implementation class to start getting a sense of the structures needed
    - Findings:
        - We need to map ProtocolContext method arguments into ProtocolEngine request objects
        - We need something to make async ProtocolEngine execution requests synchronouse
     - Decisions:
         - `ProtocolEngineContext` maps arguments to ProtocolEngine values
         - Create a `SyncClient` to map request objects and result values from the engine
         - Create a `ChildThreadTransport` to "transport" requests to the engine using `asyncio.run_coroutine_threadsafe`

2. Added a `SyncClient`, `AbstractSyncTransport`, and  `ChildThreadTransport` to a new module: `protocol_engine.clients`
    - Allows the `ProtocolEngineContext` to simply depend on `SyncClient` to the protocol engine
    - In the future, the interface can be adapted to other approaches, like multiprocessing or concurrency-via-threading
    - This part went through the most changes due to review feedback
    - See "Detailed Findings" for why I took this approach

3. Wired up `LabwareGeometry` in `ProtocolEngineContext` to go fully through the `ProtocolEngine` state
    - Needed to touch `LabwareContext` to get this PR working anyway
    - Noticed a bug with the previous engine state, which this work allowed me to squash
        - The engine was not taking into account `labware_def.cornerOffsetFromSlot`, which allows a labware's origin to be offset from the slot origin without using calibration
        - All existing definitions use `(0, 0, 0)`, but this would have been an annoying bug to track down if we ever added a def that did include an offset

### Detailed Findings

This runs us head-first into the boundary between the synchronous `ProtocolContext` and the asynchronous `ProtocolEngine`. Off the bat, I think we should agree that **challenging the assumption that ProtocolEngine should be async is _not out of scope_ for this project**.

Assuming that we _do_ want ProtocolEngine to remain async, as this PR is going to do, we have a few options for the conversion that I know of:

1. `asyncio.get_event_loop().run_until_complete(coroutine)`
    - This _may_ be the same as `asyncio.run(coroutine)`; I'm not sure
2. `asyncio.run_coroutine_threadsafe(coroutine(), loop_in_another_thread)`
    - Implies the protocol execution is running in a different thread than the engine's command execution
3. A message queue that's async on one end and synchronous on the other end
    - e.g. A zmq REQ/REP pattern 
    - Implies the protocol execution is running in a different thread or process than the engine's command execution

I'm in a place where I'm leaning towards (3) as the best long term solution, and (2) is a viable stepping stone.

- For the purposes of isolation, running Python protocol execution off the main thread is a good idea
    - Busy waits in the Python protocol don't freeze the webserver
    - Exotic stuff like [sys.settrace](https://docs.python.org/3/library/sys.html#sys.settrace) become available
    - The execution can happen free of an asynchronous context, which is [currently a problem](https://coda.io/d/Hardware-Control-Architecture_ddYiAbjtAiP)
- Running Python protocols in a separate process has a lot of potential, but would require a lot of work
- Running the Python protocol in a separate thread is something we're already doing, but we could take a more considered approach to it

So, given those points, my conclusion is: **consolidate the execution async. to sync. boundary in one explicit module** (i.e. not a proxy-style `SynchronousAdapter`), and we can iterate on the exact mechanics of that boundary (or the eventual removal of that boundary if we decide to ditch async. entirely).

## Risk assessment

N/A, code is not hooked up

### CODE REVIEW

File: api/tests/opentrons/protocols/implementations/engine/test_protocol_context.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
+import pytest
```
Comment (mcous) [inline api/tests/opentrons/protocols/implementations/engine/test_protocol_context.py:1]: Sorry for this diff! Renamed `api/tests/opentrons/protocols/implementation` to ` api/tests/opentrons/protocols/implementation[s]` to match source code, and it confused git

File: api/tests/opentrons/protocols/implementations/engine/test_protocol_context.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def test_load_labware_from_definition(
+        subject: ProtocolEngineContext) -> None:
+    with pytest.raises(NotImplementedError):
+        subject.load_labware_from_definition({}, "")
+
+
+@pytest.mark.xfail(reason="Not yet implemented", raises=NotImplementedError)
+def test_load_labware(
```
Comment (mcous) [inline api/tests/opentrons/protocols/implementations/engine/test_protocol_context.py:96]: This is the only added test

File: api/src/opentrons/protocol_engine/protocol_engine.py [hunk 3]
```diff
@@ -71,6 +71,7 @@ async def execute_command(
         command_id: str,
     ) -> Union[CompletedCommandType, FailedCommandType]:
         """Execute a command request, waiting for it to complete."""
+        print("hello!", request, command_id)
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/protocol_engine.py]: Oh no, now everybody will know I only use print statement debugging

File: api/src/opentrons/protocols/implementations/engine/sync_engine.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            request: The ProtocolEngine command request
+
+        Returns:
+            The command's result data.
+
+        Raises:
+            ProtocolEngineError: if the command execution is not successful,
+                the specific error that cause the command to fail is raised.
+        """
+        command_id = str(uuid4())
```
Comment (mcous) [inline api/src/opentrons/protocols/implementations/engine/sync_engine.py]: I think I should move this into an `IdGenerator` component or similar. It's possible we'll want to generate command IDs in Python based on something fun like the exact Python frame

**Comment by amitlissack** [comment]
I agree with your conclusion for cleanest approach if we stick with `asyncio`. 

But....

Since you mentioned **challenging the assumption that ProtocolEngine should be async is not out of scope for this project**, I cannot help but feel a little icky inside that we to jump through hoops (adapters, thread managers, etc) just to accommodate `asyncio`. I just don't know what `asyncio` buys us.   

It's not just a `ProtocolEngine` problem. It's a hardware controller / driver problem. 

I know I've said this before... but I am bugged by this question: 
`If the Protocol API does not use `asyncio` and running protocols is what the robot is built for, why do we use `asyncio` at all?`

To me it seems that we haven't done `asyncio` well so far. Many of our issues are because we're on this fence.

This isn't really a prescription or suggestion. Just worth thinking about. We don't have to use `asyncio`.

--- 

All that being said, as we have live protocols and an isolated JSON runner (which can be purely `asyncio` - all adapterless). This isolates the async<->sync to just the protocol context. If we created a separate process/thread just to run the python protocol communicating to `ProtocolEngine` over zmq (or something) that would be pretty clean. 

---

Can we consider the performance implications of this adapter on simulation?

File: api/src/opentrons/protocol_engine/types.py [hunk 5]
```diff
@@ -43,3 +44,13 @@ class DeckLocation(BaseModel):
...
     labware_id: str
     well_name: str
+
+
+@final
+@dataclass(frozen=True)
+class Dimensions:
+    """Dimensions of an object in deck-space."""
+
+    x_dimension: float
```
Comment (amitlissack) [inline api/src/opentrons/protocol_engine/types.py]: Minor minor point. I know it's a holdover from the geometry, but is `_dimension` suffix necessary?

File: api/src/opentrons/protocols/geometry/labware_geometry.py [hunk 6]
```diff
@@ -1,8 +1,39 @@
+from abc import ABC, abstractmethod
+
 from opentrons.types import Location, Point
 from opentrons_shared_data.labware.dev_types import LabwareDefinition
 
 
-class LabwareGeometry:
+class AbstractLabwareGeometry(ABC):
```
Comment (amitlissack) [inline api/src/opentrons/protocols/geometry/labware_geometry.py:7]: Very nice!

**Review by amitlissack — APPROVED** [review]
Looks really good!

File: api/src/opentrons/protocols/implementations/engine/protocol_context.py [hunk 7]
```diff
@@ -16,15 +18,21 @@
...
 
+from .labware_context import LabwareContext
+
 
 class ProtocolEngineContext(ProtocolContextInterface):
     """ProtocolContextInterface that interacts with Protocol Engine"""
 
-    _protocol_engine: ProtocolEngine
+    _client: ProtocolEngineClient
+    _state_view: StateView
```
Comment (mcous) [inline api/src/opentrons/protocols/implementations/engine/protocol_context.py]: Remove this line

**Review by sanni-t — APPROVED** [review]
🧫

**Pull request merged**